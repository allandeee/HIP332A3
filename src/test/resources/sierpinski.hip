// Hipster cellular automaton language.
//
// Simple one dimensional automaton, presented as a 2-dimensional
// progression down the screen.
//
// This example shows that we could do with simplifying things by
// adding a few more constructs to the Hipster language. How about
// a switch statement.
//
// Â© 2017, Dominic Verity, Macquarie University, All rights reserved.
// 
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

// Constants giving the sizes of the dimensions.
int width = 402;
int height = 200;

// Constants determining the specific one-dimensional automaton to
// execute. These choices give the Sierpinski gasket.
boolean a0 = false;
boolean a1 = true;
boolean a2 = false;
boolean a3 = true;
boolean a4 = true;
boolean a5 = false;
boolean a6 = true;
boolean a7 = false;

// Automaton dimension
dimension(width + 1, height + 1);

// The neighbours of a cell are the three cells immediately above it.
neighbourhood L = [-1,1], C = [0,1], R = [1,1];

// A cell can be either alive or dead.
state {
  boolean alive = false;
}

// Updater. There are 8 possibilities for the conbination of states
// of the three neighbours of the current cell. For each such combination
// we set the state of the current cell in the next generation to the
// value of one of the constants a0...a7.
updater {
  if L:alive then {
    if C:alive then {
       if R:alive then {
         alive = a7;
       } else {
         alive = a6;
       }
    } else {
       if R:alive then {
         alive = a5; 
       } else {
         alive = a4;
       }
    }
  } else {
    if C:alive then {
       if R:alive then {
         alive = a3;
       } else {
         alive = a2;
       }
    } else {
       if R:alive then {
         alive = a1;
       } else {
         alive = a0;
       }
    }
  }
}

// Colour constants
int white = 0xFFFFFF;
int cyan = 0x00FFFF; 

// Dead cells displayed in white, live cells in cyan.
mapper {
  if alive then return(cyan); else return(white);
}

// Initialiser, set a single point at the top of the screen,
// but just outside of the world boundary, to alive.
// Since this cell is outside the boundary it won't get changed.
// But when this initialiser is selected nothing will change on
// on the screen.
initialiser select_me {
  cell [width/2, height] {
    C:alive = true;
  }
}
