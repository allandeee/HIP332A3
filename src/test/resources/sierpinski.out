package org.bitbucket.dominicverity.cellsim;

class CellularAutomaton extends BaseAutomaton {

    private static final int width$5 = 402;
    private static final int height$6 = 200;
    private static final boolean a0$7 = false;
    private static final boolean a1$8 = true;
    private static final boolean a2$9 = false;
    private static final boolean a3$10 = true;
    private static final boolean a4$11 = true;
    private static final boolean a5$12 = false;
    private static final boolean a6$13 = true;
    private static final boolean a7$14 = false;
    private static final int white$19 = 16777215;
    private static final int cyan$20 = 65535;

    class Cell extends BaseCell {

        Cell _other$2;
        Cell _theother$3;
        Cell L$15;
        Cell C$16;
        Cell R$17;
        boolean alive$18;

        Cell(Point _loc$1) {
            super(_loc$1);
            _other$2 = this;
            resetCellState();
        }

        void initNeighbours() {
            _theother$3 = getOther();
            L$15 = getCellRel(-1, 1);
            C$16 = getCellRel(0, 1);
            R$17 = getCellRel(1, 1);
        }

        synchronized void updater() {
            _other$2.alive$18 = alive$18;
            if (L$15.alive$18) {
                if (C$16.alive$18) {
                    if (R$17.alive$18) {
                        _other$2.alive$18 = a7$14;
                    }
                    else {
                        _other$2.alive$18 = a6$13;
                    }
                }
                else {
                    if (R$17.alive$18) {
                        _other$2.alive$18 = a5$12;
                    }
                    else {
                        _other$2.alive$18 = a4$11;
                    }
                }
            }
            else {
                if (C$16.alive$18) {
                    if (R$17.alive$18) {
                        _other$2.alive$18 = a3$10;
                    }
                    else {
                        _other$2.alive$18 = a2$9;
                    }
                }
                else {
                    if (R$17.alive$18) {
                        _other$2.alive$18 = a1$8;
                    }
                    else {
                        _other$2.alive$18 = a0$7;
                    }
                }
            }
        }

        synchronized int mapper() {
            if (alive$18)
                return cyan$20;
            else
                return white$19;
        }

        synchronized void resetCellState() {
            alive$18 = false;
        }

        Cell getCellRel(int... _off$4) {
            return (Cell)getCell(_off$4);
        }

        Cell getOther() {
            return (Cell)getNextGen();
        }

        void otherToSelf() {
            _other$2 = this;
        }

        void otherToOther() {
            _other$2 = _theother$3;
        }
    }

    CellularAutomaton() {
        super();
        initDims(width$5 + 1, height$6 + 1);
        initCycles(false, false);
        initWorld();
        addInitialiser("select_me", (() -> { select_me$21(); }));
    }

    void select_me$21() {
        {
            Cell _cell$22 = getCellAbs(width$5 / 2, height$6);
            {
                _cell$22.C$16._other$2.alive$18 = true;
            }
        }
    }

    BaseCell makeCell(Point _loc$1) {
        return new Cell(_loc$1);
    }

    Cell getCellAbs(int... _loc$1) {
        return (Cell)getCell(_loc$1);
    }
}
